### PR TITLE
RavenDB-20265 Optimizing Corax runtime & disk usage

### PR DESCRIPTION
**Pull request by ayende**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20265 

### Additional description

* Prevent degenerate case where Corax would consume more disk space than the source data because of limits in big posting lists.
* Moved small posting lists to use PFor encoding
* Use VxSort for speeding up large indexing batches
* More tests

### Type of change

- Bug fix

### How risky is the change?

- High

### Backward compatibility

- Breaking change

### CODE REVIEW

File: src/Voron/Data/PostingLists/PostingList.cs [hunk 1]
```diff
@@ -131,12 +132,14 @@ private void ResizeCursorState()
             return state;
         }
 
+        [Conditional("DEBUG")]
         public void Verify()
```
Comment (redknightlois) [inline src/Voron/Data/PostingLists/PostingList.cs:136]: Is this correct? If we rely on calling Verify in tests this change would essentially nullify the test.

Comment (ayende) [reply src/Voron/Data/PostingLists/PostingList.cs:136]: We are never calling this right now.
This is debug code only, I used it to verify that the structure of the tree is sound.
We are testing the actual semantics by reading the data, which does the same better.
This is at a lower level.

**Comment by arekpalinski** [comment]
1. Many tests have failed (https://jenkins.hibernatingrhinos.com/job/PR_Tests/5778/ - but it didn't look that it was related to changed made here) so I rerun the tests again.
2. Is there any reason why not to apply it to `v5.4` branch?

**Comment by arekpalinski** [comment]
Many failing tests: https://jenkins.hibernatingrhinos.com/job/PR_Tests/5784/

**Comment by redknightlois** [comment]
This PR is still not ready AFAIK, we have been discussing some work yesterday with @ayende...

**Comment by ayende** [comment]
This should now be viable to actual test.
For some reason, Lucene tests failed on MacOS, no idea why

**Comment by redknightlois** [comment]
LGTM

**Review by redknightlois — APPROVED** [review]

**Comment by arekpalinski** [comment]
1. Please check failing tests: https://jenkins.hibernatingrhinos.com/job/PR_Tests/5849

2. Let's introduce those changes to v5.4 branch. It will be way easier to deal with 5.4 to 6.0 merges.

**Comment by ayende** [comment]
Closing this, will open against 5.4 only with just the indexing portions

**Pull request closed without merging**